kotlinx-coroutines-core/common/src/flow/SharedFlow.kt

```diff
@@ -0,0 +1,652 @@
+ * For example, the following class encapsulates an event bus that distributes events to all subscribers
+ * in _rendezvous_ manner, suspending until all subscribers process each event:
+ *
+ * ```
+ * class EventBus {
+ *     private val _events = MutableSharedFlow<Event>(0) // private mutable shared flow
+ *     val events get() = _events.asSharedFlow() // publicly exposed as read-only shared flow
+ *
+ *     suspend fun produceEvent(event: Event) {
+ *         _events.emit(event) // suspends until all subscribers receive it
+ *     }
+ * }
+ * ```
+ *
+ * As an alternative to the above usage with `MutableSharedFlow(...)` constructor function,
+ * any _cold_ [Flow] can be converted to a shared flow using [shareIn] operator.
+ *
+ * There is a specialized implementation of shared flow for a case where the most recent state value needs
+ * to be shared. See [StateFlow] for details.
+ *
+ * ### Replay cache and buffer
+ *
+ * A shared flow keeps a specific number of the most recent values in its _replay cache_. Every new subscribers first
+ * gets the values from the replay cache and then gets new emitted values. The maximal size of the replay cache is
+ * specified when the shared flow is created by the `replay` parameter. A snapshot of the current replay cache
+ * is available via [replayCache] property and it can be reset with [MutableSharedFlow.resetReplayCache] function.
+ *
+ * A replay cache provides buffer for emissions to the shared flow. Buffer space allows slow subscribers to
+ * get values from the buffer without suspending emitters. The buffer space determines how much slow subscribers
+ * can lag from the fast ones. When creating a shared flow additional buffer capacity beyond replay can be reserved
+ * using `extraBufferCapacity` parameter.
+ * 
+ * A shared flow with a buffer can be configured to avoid suspension of emitters on buffer overflow using
+ * `onBufferOverflow` parameter, which is equal to one of the entries of [BufferOverflow] enum. When a strategy other
+ * than [SUSPENDED][BufferOverflow.SUSPEND] is configured emissions to the shared flow never suspend.
+ *
+ * ### SharedFlow vs BroadcastChannel
+ *
+ * Conceptually shared flow is similar to [BroadcastChannel][BroadcastChannel]
+ * and is designed to completely replace `BroadcastChannel` in the future.
+ * It has the following important differences:
+ *
+ * * `SharedFlow` is simpler, because it does not have to implement all the [Channel] APIs, which allows
+ *    for faster and simpler implementation.
+ * * `SharedFlow` supports configurable replay and buffer overflow strategy.
+ * * `SharedFlow` has a clear separation into a read-only `SharedFlow` interface and a [MutableSharedFlow].
+ * * `SharedFlow` cannot be closed like `BroadcastChannel` and can never represent a failure.
+ *    All errors and completion signals shall be explicitly _materialized_ if needed.
```

Comment:
Maybe add a short description of materialization, or link to somewhere that provides an explanation.

Reply:
We actually plan to provide out-of-the box materialization operators a little bit later in the future (see #2092) and this piece of documentation will simply link to them, so writing a detailed example of how to do it manually will be somewhat wasted effort.